### ISSUE TITLE
ANR on specific device Moto G.

### ISSUE THREAD

**Issue by yingchun1** [smartdevicelink/sdl_java_suite#829]
### Bug Report
ANR of our end user.

##### Reproduction Steps
1. Not sure the exact step but our user report ANR most in a specific device Moto G.

##### Expected Behavior
it should not ANR when starting sdlrouterservice

##### Observed Behavior
meet ANR

##### OS & Version Information
* Android Version: 7.0 & 7.1
* SDL Android Version: 4.6.0

Most ANR device info:
![image](https://user-images.githubusercontent.com/20849782/43299733-0f2ea36a-918e-11e8-8bb7-e9562e7f31ef.png)

Some log of the ANR:

```
`"main" prio=5 tid=1 Native
  | group="main" sCount=1 dsCount=0 obj=0x73e52730 self=0xb2685400
  | sysTid=25675 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0xb53da534
  | state=S schedstat=( 323897979 8251093 355 ) utm=25 stm=6 core=3 HZ=100
  | stack=0xbe6ff000-0xbe701000 stackSize=8MB
  | held mutexes=
  #00  pc 00000000000495d8  /system/lib/libc.so (recvmsg+8)
  #01  pc 0000000000094ef5  /system/lib/libandroid_runtime.so (???)
  #02  pc 00000000000952c1  /system/lib/libandroid_runtime.so (???)
  #03  pc 000000000052eee9  /system/framework/arm/boot-framework.oat (Java_android_net_LocalSocketImpl_readba_1native___3BIILjava_io_FileDescriptor_2+140)
  at android.net.LocalSocketImpl.readba_native (Native method)
  at android.net.LocalSocketImpl.-wrap1 (LocalSocketImpl.java)
  at android.net.LocalSocketImpl$SocketInputStream.read (LocalSocketImpl.java:110)
- locked <0x0c5425c4> (a java.lang.Object)
  at android.bluetooth.BluetoothSocket.readAll (BluetoothSocket.java:736)
  at android.bluetooth.BluetoothSocket.readInt (BluetoothSocket.java:750)
  at android.bluetooth.BluetoothSocket.bindListen (BluetoothSocket.java:427)
  at android.bluetooth.BluetoothAdapter.createNewRfcommSocketAndRecord (BluetoothAdapter.java:1708)
  at android.bluetooth.BluetoothAdapter.listenUsingRfcommWithServiceRecord (BluetoothAdapter.java:1628)
  at com.smartdevicelink.transport.MultiplexBluetoothTransport$AcceptThread.<init> (SourceFile:391)
- locked <0x0cedfcad> (a java.lang.Object)
  at com.smartdevicelink.transport.MultiplexBluetoothTransport.start (SourceFile:195)
  at com.smartdevicelink.transport.SdlRouterService.initBluetoothSerialService (SourceFile:1230)
- locked <0x0bb22a71> (a com.cisco.webex.meetings.SdlRouterService)
  at com.smartdevicelink.transport.SdlRouterService.startUpSequence (SourceFile:875)
  at com.smartdevicelink.transport.SdlRouterService.onCreate (SourceFile:822)
  at android.app.ActivityThread.handleCreateService (ActivityThread.java:3212)
  at android.app.ActivityThread.-wrap5 (ActivityThread.java)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1580)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:154)
  at android.app.ActivityThread.main (ActivityThread.java:6165)
  at java.lang.reflect.Method.invoke! (Native method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:888)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:778)
"Jit thread pool worker thread 0" tid=2 Native 
"Jit thread pool worker thread 0" prio=5 tid=2 Native (still starting up)
  | group="" sCount=1 dsCount=0 obj=0x0 self=0xaad0d000
  | sysTid=25680 nice=9 cgrp=bg_non_interactive sched=0/0 handle=0xb1da3920
  | state=S schedstat=( 349583 172032 2 ) utm=0 stm=0 core=2 HZ=100
  | stack=0xb1ca5000-0xb1ca7000 stackSize=1022KB
  | held mutexes=
  #00  pc 00000000000174e4  /system/lib/libc.so (syscall+28)
  #01  pc 00000000000b6fc9  /system/lib/libart.so (_ZN3art17ConditionVariable16WaitHoldingLocksEPNS_6ThreadE+92)
  #02  pc 0000000000345909  /system/lib/libart.so (_ZN3art10ThreadPool7GetTaskEPNS_6ThreadE+160)
  #03  pc 0000000000345133  /system/lib/libart.so (_ZN3art16ThreadPoolWorker3RunEv+62)
  #04  pc 0000000000344c61  /system/lib/libart.so (_ZN3art16ThreadPoolWorker8CallbackEPv+64)
  #05  pc 0000000000046ee3  /system/lib/libc.so (_ZL15__pthread_startPv+22)
  #06  pc 0000000000019afd  /system/lib/libc.so (__start_thread+6)
"FinalizerDaemon" tid=4 Waiting 
"FinalizerDaemon" daemon prio=5 tid=4 Waiting
  | group="system" sCount=1 dsCount=0 obj=0x2ac09280 self=0xa5a17600
  | sysTid=25683 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0xb1a9e920
  | state=S schedstat=( 1376250 176145 11 ) utm=0 stm=0 core=3 HZ=100
  | stack=0xb199c000-0xb199e000 stackSize=1038KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x0f1f7be2> (a java.lang.Object)
  at java.lang.Object.wait (Object.java:407)
  at java.lang.ref.ReferenceQueue.remove (ReferenceQueue.java:188)
- locked <0x0f1f7be2> (a java.lang.Object)
  at java.lang.ref.ReferenceQueue.remove (ReferenceQueue.java:209)
  at java.lang.Daemons$FinalizerDaemon.run (Daemons.java:204)
  at java.lang.Thread.run (Thread.java:761)
"ReferenceQueueDaemon" tid=5 Waiting 
"ReferenceQueueDaemon" daemon prio=5 tid=5 Waiting
  | group="system" sCount=1 dsCount=0 obj=0x2ac09310 self=0xa5a17100
  | sysTid=25682 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0xb1ba3920
  | state=S schedstat=( 1174740 490157 14 ) utm=0 stm=0 core=0 HZ=100
  | stack=0xb1aa1000-0xb1aa3000 stackSize=1038KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x0af2cb73> (a java.lang.Class<java.lang.ref.ReferenceQueue>)
  at java.lang.Daemons$ReferenceQueueDaemon.run (Daemons.java:150)
- locked <0x0af2cb73> (a java.lang.Class<java.lang.ref.ReferenceQueue>)
  at java.lang.Thread.run (Thread.java:761)
"FinalizerWatchdogDaemon" tid=6 Waiting 
"FinalizerWatchdogDaemon" daemon prio=5 tid=6 Waiting
  | group="system" sCount=1 dsCount=0 obj=0x2ac093a0 self=0xa5a17b00
  | sysTid=25684 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0xb1999920
  | state=S schedstat=( 539323 783645 15 ) utm=0 stm=0 core=3 HZ=100
  | stack=0xb1897000-0xb1899000 stackSize=1038KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x014fda30> (a java.lang.Daemons$FinalizerWatchdogDaemon)
  at java.lang.Daemons$FinalizerWatchdogDaemon.sleepUntilNeeded (Daemons.java:269)
- locked <0x014fda30> (a java.lang.Daemons$FinalizerWatchdogDaemon)
  at java.lang.Daemons$FinalizerWatchdogDaemon.run (Daemons.java:249)
  at java.lang.Thread.run (Thread.java:761)
"Binder:25675_1" tid=8 Native 
"Binder:25675_1" prio=5 tid=8 Native
  | group="main" sCount=1 dsCount=0 obj=0x2ac094c0 self=0xaad0e400
  | sysTid=25688 nice=0 cgrp=default sched=0/0 handle=0xb1691920
  | state=S schedstat=( 5083491 1369113 20 ) utm=0 stm=0 core=5 HZ=100
  | stack=0xb1595000-0xb1597000 stackSize=1014KB
  | held mutexes=
  #00  pc 0000000000048498  /system/lib/libc.so (__ioctl+8)
  #01  pc 000000000001aaff  /system/lib/libc.so (ioctl+38)
  #02  pc 000000000003cd65  /system/lib/libbinder.so (_ZN7android14IPCThreadState14talkWithDriverEb+168)
  #03  pc 000000000003ce5d  /system/lib/libbinder.so (_ZN7android14IPCThreadState20getAndExecuteCommandEv+8)
  #04  pc 000000000003d3e3  /system/lib/libbinder.so (_ZN7android14IPCThreadState14joinThreadPoolEb+46)
  #05  pc 000000000004f819  /system/lib/libbinder.so (???)
  #06  pc 000000000000e435  /system/lib/libutils.so (_ZN7android6Thread11_threadLoopEPv+140)
  #07  pc 0000000000066ab5  /system/lib/libandroid_runtime.so (_ZN7android14AndroidRuntime15javaThreadShellEPv+80)
  #08  pc 0000000000046ee3  /system/lib/libc.so (_ZL15__pthread_startPv+22)
  #09  pc 0000000000019afd  /system/lib/libc.so (__start_thread+6)
"Binder:25675_2" tid=9 Native 
"Binder:25675_2" prio=5 tid=9 Native
  | group="main" sCount=1 dsCount=0 obj=0x2ac09550 self=0xb2686300
  | sysTid=25689 nice=0 cgrp=default sched=0/0 handle=0xb1592920
  | state=S schedstat=( 4895779 8285158 19 ) utm=0 stm=0 core=3 HZ=100
  | stack=0xb1496000-0xb1498000 stackSize=1014KB
  | held mutexes=
  #00  pc 0000000000048498  /system/lib/libc.so (__ioctl+8)
  #01  pc 000000000001aaff  /system/lib/libc.so (ioctl+38)
  #02  pc 000000000003cd65  /system/lib/libbinder.so (_ZN7android14IPCThreadState14talkWithDriverEb+168)
  #03  pc 000000000003ce5d  /system/lib/libbinder.so (_ZN7android14IPCThreadState20getAndExecuteCommandEv+8)
  #04  pc 000000000003d3e3  /system/lib/libbinder.so (_ZN7android14IPCThreadState14joinThreadPoolEb+46)
  #05  pc 000000000004f819  /system/lib/libbinder.so (???)
  #06  pc 000000000000e435  /system/lib/libutils.so (_ZN7android6Thread11_threadLoopEPv+140)
  #07  pc 0000000000066ab5  /system/lib/libandroid_runtime.so (_ZN7android14AndroidRuntime15javaThreadShellEPv+80)
  #08  pc 0000000000046ee3  /system/lib/libc.so (_ZL15__pthread_startPv+22)
  #09  pc 0000000000019afd  /system/lib/libc.so (__start_thread+6)
"Profile Saver" tid=10 Native 
"Profile Saver" daemon prio=5 tid=10 Native
  | group="system" sCount=1 dsCount=0 obj=0x2ac095e0 self=0xaad0f800
  | sysTid=25690 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0xaafcd920
  | state=S schedstat=( 14377294 874322 9 ) utm=1 stm=0 core=0 HZ=100
  | stack=0xaaed1000-0xaaed3000 stackSize=1014KB
  | held mutexes=
  #00  pc 00000000000174e4  /system/lib/libc.so (syscall+28)
  #01  pc 00000000000b6fc9  /system/lib/libart.so (_ZN3art17ConditionVariable16WaitHoldingLocksEPNS_6ThreadE+92)
  #02  pc 000000000025b77d  /system/lib/libart.so (_ZN3art12ProfileSaver3RunEv+296)
  #03  pc 000000000025cab9  /system/lib/libart.so (_ZN3art12ProfileSaver21RunProfileSaverThreadEPv+52)
  #04  pc 0000000000046ee3  /system/lib/libc.so (_ZL15__pthread_startPv+22)
  #05  pc 0000000000019afd  /system/lib/libc.so (__start_thread+6)
"preMeetingProcessThread" tid=13 Native 
"preMeetingProcessThread" prio=5 tid=13 Native
  | group="main" sCount=1 dsCount=0 obj=0x2ac09790 self=0xa8209400
  | sysTid=25694 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0x98745920
  | state=S schedstat=( 323490 102188 6 ) utm=0 stm=0 core=1 HZ=100
  | stack=0x98643000-0x98645000 stackSize=1038KB
  | held mutexes=
  #00  pc 0000000000048358  /system/lib/libc.so (__epoll_pwait+20)
  #01  pc 0000000000019ead  /system/lib/libc.so (epoll_pwait+60)
  #02  pc 0000000000019edd  /system/lib/libc.so (epoll_wait+12)
  #03  pc 0000000000011def  /system/lib/libutils.so (_ZN7android6Looper9pollInnerEi+118)
  #04  pc 0000000000011ceb  /system/lib/libutils.so (_ZN7android6Looper8pollOnceEiPiS1_PPv+26)
  #05  pc 0000000000092c79  /system/lib/libandroid_runtime.so (_ZN7android18NativeMessageQueue8pollOnceEP7_JNIEnvP8_jobjecti+22)
  #06  pc 00000000006427f5  /system/framework/arm/boot-framework.oat (Java_android_os_MessageQueue_nativePollOnce__JI+96)
  at android.os.MessageQueue.nativePollOnce (Native method)
  at android.os.MessageQueue.next (MessageQueue.java:323)
  at android.os.Looper.loop (Looper.java:136)
  at android.os.HandlerThread.run (HandlerThread.java:61)
"pool-1-thread-1" tid=14 Waiting 
"pool-1-thread-1" prio=5 tid=14 Waiting
  | group="main" sCount=1 dsCount=0 obj=0x2ac09820 self=0xaad10c00
  | sysTid=25695 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0x98640920
  | state=S schedstat=( 392137 144010 6 ) utm=0 stm=0 core=3 HZ=100
  | stack=0x9853e000-0x98540000 stackSize=1038KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x049ab6a9> (a java.lang.Object)
  at java.lang.Thread.parkFor$ (Thread.java:2127)
- locked <0x049ab6a9> (a java.lang.Object)
  at sun.misc.Unsafe.park (Unsafe.java:325)
  at java.util.concurrent.locks.LockSupport.park (LockSupport.java:161)
  at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await (AbstractQueuedSynchronizer.java:2035)
  at java.util.concurrent.LinkedBlockingQueue.take (LinkedBlockingQueue.java:413)
  at java.util.concurrent.ThreadPoolExecutor.getTask (ThreadPoolExecutor.java:1058)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1118)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:607)
  at java.lang.Thread.run (Thread.java:761)
"DBThread" tid=15 Waiting 
"DBThread" prio=5 tid=15 Waiting
  | group="main" sCount=1 dsCount=0 obj=0x2ac098b0 self=0xa820ad00
  | sysTid=25697 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0x8e5af920
  | state=S schedstat=( 525571 4757448 16 ) utm=0 stm=0 core=2 HZ=100
  | stack=0x8e4ad000-0x8e4af000 stackSize=1038KB
  | held mutexes=
  at java.lang.Object.wait! (Native method)
- waiting on <0x093d742e> (a java.util.Vector)
  at ayp.run (SourceFile:432)
- locked <0x093d742e> (a java.util.Vector)
  at java.lang.Thread.run (Thread.java:761)
"Signal Catcher" tid=3 Runnable 
"Signal Catcher" daemon prio=5 tid=3 Runnable
  | group="system" sCount=0 dsCount=0 obj=0x2ac091f0 self=0xa5a15d00
  | sysTid=25681 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0xb1ca2920
  | state=R schedstat=( 16365053 2315261 23 ) utm=0 stm=0 core=2 HZ=100
  | stack=0xb1ba6000-0xb1ba8000 stackSize=1014KB
  | held mutexes= "mutator lock"(shared held)
  #00  pc 0000000000351ba1  /system/lib/libart.so (_ZN3art15DumpNativeStackERNSt3__113basic_ostreamIcNS0_11char_traitsIcEEEEiP12BacktraceMapPKcPNS_9ArtMethodEPv+128)
  #01  pc 00000000003323a5  /system/lib/libart.so (_ZNK3art6Thread9DumpStackERNSt3__113basic_ostreamIcNS1_11char_traitsIcEEEEbP12BacktraceMap+308)
  #02  pc 00000000003443c1  /system/lib/libart.so (_ZN3art14DumpCheckpoint3RunEPNS_6ThreadE+620)
  #03  pc 000000000033e3cb  /system/lib/libart.so (_ZN3art10ThreadList13RunCheckpointEPNS_7ClosureE+330)
  #04  pc 000000000033e0c7  /system/lib/libart.so (_ZN3art10ThreadList4DumpERNSt3__113basic_ostreamIcNS1_11char_traitsIcEEEEb+586)
  #05  pc 000000000033de1f  /system/lib/libart.so (_ZN3art10ThreadList14DumpForSigQuitERNSt3__113basic_ostreamIcNS1_11char_traitsIcEEEE+574)
  #06  pc 00000000003232b5  /system/lib/libart.so (_ZN3art7Runtime14DumpForSigQuitERNSt3__113basic_ostreamIcNS1_11char_traitsIcEEEE+124)
  #07  pc 0000000000327d3b  /system/lib/libart.so (_ZN3art13SignalCatcher13HandleSigQuitEv+1394)
  #08  pc 0000000000326ead  /system/lib/libart.so (_ZN3art13SignalCatcher3RunEPv+352)
  #09  pc 0000000000046ee3  /system/lib/libc.so (_ZL15__pthread_startPv+22)
  #10  pc 0000000000019afd  /system/lib/libc.so (__start_thread+6)
"HeapTaskDaemon" tid=7 Blocked 
"HeapTaskDaemon" daemon prio=5 tid=7 Blocked
  | group="system" sCount=1 dsCount=0 obj=0x2ac09430 self=0xb2685e00
  | sysTid=25686 nice=0 cgrp=bg_non_interactive sched=0/0 handle=0xb1894920
  | state=S schedstat=( 88177653 3827031 35 ) utm=7 stm=0 core=2 HZ=100
  | stack=0xb1792000-0xb1794000 stackSize=1038KB
  | held mutexes=
  #00  pc 00000000000174e4  /system/lib/libc.so (syscall+28)
  #01  pc 00000000000b6fc9  /system/lib/libart.so (_ZN3art17ConditionVariable16WaitHoldingLocksEPNS_6ThreadE+92)
  #02  pc 00000000001b51e5  /system/lib/libart.so (_ZN3art2gc13TaskProcessor7GetTaskEPNS_6ThreadE+288)
  #03  pc 00000000001b585d  /system/lib/libart.so (_ZN3art2gc13TaskProcessor11RunAllTasksEPNS_6ThreadE+44)
  #04  pc 0000000000161577  /system/framework/arm/boot-core-libart.oat (Java_dalvik_system_VMRuntime_runHeapTasks__+74)
  at dalvik.system.VMRuntime.runHeapTasks (Native method)
- waiting to lock an unknown object
  at java.lang.Daemons$HeapTaskDaemon.run (Daemons.java:433)
  at java.lang.Thread.run (Thread.java:761)`

```

**Comment by joeygrover** [comment]
This looks to be an issue with the specific Bluetooth stack implemented on the Moto G as the log shows it made it down to the socket creation from the native layer. This isn't a bug specific to the project but should be addressed. 

```
  at android.net.LocalSocketImpl.readba_native (Native method)
  at android.net.LocalSocketImpl.-wrap1 (LocalSocketImpl.java)
  at android.net.LocalSocketImpl$SocketInputStream.read (LocalSocketImpl.java:110)
- locked <0x0c5425c4> (a java.lang.Object)
  at android.bluetooth.BluetoothSocket.readAll (BluetoothSocket.java:736)
  at android.bluetooth.BluetoothSocket.readInt (BluetoothSocket.java:750)
  at android.bluetooth.BluetoothSocket.bindListen (BluetoothSocket.java:427)
  at android.bluetooth.BluetoothAdapter.createNewRfcommSocketAndRecord (BluetoothAdapter.java:1708)
  at android.bluetooth.BluetoothAdapter.listenUsingRfcommWithServiceRecord (BluetoothAdapter.java:1628)
  at com.smartdevicelink.transport.MultiplexBluetoothTransport$AcceptThread.<init> (SourceFile:391)
- locked <0x0cedfcad> (a java.lang.Object)
  at com.smartdevicelink.transport.MultiplexBluetoothTransport.start (SourceFile:195)
```

**Comment by yingchun1** [comment]
if So, do you have plan on working on this "device" issue? Top 1 ANR happend on this device.

**Comment by joeygrover** [comment]
This is the first reported issue of this type so you will have to give us some time to investigate solutions to work around the device specific behavior.

**Comment by yingchun1** [comment]
ok. Thank you~~

**Comment by Tnnnguyen** [comment]
@yingchun1 

I was able to create an anr only one time, see log and screenshots. The exception that resulted in the anr appeared to be an issue that was fixed with v4.6.2 [here](https://github.com/smartdevicelink/sdl_android/releases/tag/4.6.2). The library ran into an illegal state exception, the WebEx app showed the dialog with code 6 error, then the Android OS showed its anr dialog.

Although at this time, I couldn't get the exact system log, my hypothesis is that your reported anr's are related to this anr as they are thrown during router discovery and connecting.  The fix for v4.6.2 was an attempt to handle the illegal state exception 4.6.0 (in SdlBroadcastReceiver).  This particular issue was not found in 4.5.0 for the WebEx app as confirmed.  Furthermore I have looked at the changes between 4.5.0 and 4.6.0 in the 2 other main classes involved, SdlRouterService and MultiplexBluetoothTransport, and found that these changes are minimal.  

This anr is extremely intermittent, I don't have the exact steps to reproduce yet.  however, while we are working towards verifying the exact issue it wouldn't hurt to pull in the fix for v4.6.2 for a new release.

![webex_dialog](https://user-images.githubusercontent.com/8460594/43970819-ac303840-9c83-11e8-83e6-e0afdcf5e5eb.png)


![google_anr](https://user-images.githubusercontent.com/8460594/43970843-c0948390-9c83-11e8-936f-45c9b3a4a26f.png)

[WebEx_ANR_log.txt](https://github.com/smartdevicelink/sdl_android/files/2278741/WebEx_ANR_log.txt)
